### PR TITLE
Fix ghosting in KaHIP and ParMetis

### PR DESCRIPTION
**Pull request by IgorBaratta**

### CODE REVIEW

File: cpp/dolfinx/graph/KaHIP.cpp [hunk 1]
```diff
@@ -20,9 +20,8 @@ using namespace dolfinx;
 
 graph::AdjacencyList<std::int32_t> dolfinx::graph::KaHIP::partition(
     MPI_Comm mpi_comm, int nparts,
-    const graph::AdjacencyList<unsigned long long>& adj_graph)
+    const graph::AdjacencyList<unsigned long long>& adj_graph, bool ghosting)
```
Comment (garth-wells) [inline cpp/dolfinx/graph/KaHIP.cpp]: Could we used fixed-width ints in place of `unsigned long long`?

File: cpp/dolfinx/graph/KaHIP.cpp [hunk 2]
```diff
@@ -66,113 +65,111 @@ graph::AdjacencyList<std::int32_t> dolfinx::graph::KaHIP::partition(
...
+  // Create a map of local nodes to their additional destination processes,
+  // due to ghosting. If no ghosting, this will remain empty.
+  std::map<std::int32_t, std::set<std::int32_t>> local_node_to_dests;
+  if (ghosting)
   {
-    for (int j = 0; j < adj_graph.num_links(i); ++j)
+    common::Timer timer2("Compute graph halo data (KaHIP)");
+
+    // Work out halo cells for current division of dual graph
+    const std::int32_t* xadj = adj_graph.offsets().data();
```
Comment (garth-wells) [inline cpp/dolfinx/graph/KaHIP.cpp]: Can this be moved closer to where it's used?

File: cpp/dolfinx/graph/KaHIP.cpp [hunk 3]
```diff
@@ -66,113 +65,111 @@ graph::AdjacencyList<std::int32_t> dolfinx::graph::KaHIP::partition(
...
-        proc_other = part[other_cell - elm_begin];
+      cell_ownership[*p] = *(p + 1);
+    }
 
-      if (proc_this != proc_other)
+    // Generate mapping for where new boundary cells need to be sent
+    for (std::int32_t i = 0; i < ncells; i++)
+    {
+      const std::int32_t proc_this = part[i];
+      for (std::int32_t j = xadj[i]; j < xadj[i + 1]; ++j)
```
Comment (garth-wells) [inline cpp/dolfinx/graph/KaHIP.cpp]: Is `xadj` a raw pointer? Can we use it via Eigen?

File: cpp/dolfinx/graph/KaHIP.cpp (hunk 1 above)
Comment (chrisrichardson) [reply cpp/dolfinx/graph/KaHIP.cpp]: It is written into KaHIP. Not sure what are the implications for maintaining compatibility on all systems.

File: cpp/dolfinx/graph/KaHIP.cpp (hunk 3 above)
Comment (chrisrichardson) [reply cpp/dolfinx/graph/KaHIP.cpp]: Yes, there is no need to use it raw here.

File: cpp/dolfinx/graph/KaHIP.cpp (hunk 2 above)
Comment (IgorBaratta) [reply cpp/dolfinx/graph/KaHIP.cpp]: Move down and changed to `Eigen::Map`

File: cpp/dolfinx/graph/KaHIP.cpp (hunk 3 above)
Comment (IgorBaratta) [reply cpp/dolfinx/graph/KaHIP.cpp]: Using `Eigen::Map` now.

**Comment by chrisrichardson** [comment]
Does it need map? Isn't it already just an eigen array?

**Comment by IgorBaratta** [comment]
> Does it need map? Isn't it already just an eigen array?

No need at all. Just changed this.

**Pull request merged**